src/server/plugins/engine/pageControllers/FileUploadPageController.ts

```diff
@@ -0,0 +1,315 @@
+    return async (request: Request, h: ResponseToolkit) => {
+      await this.refreshUpload(request)
+
+      return super.makePostRouteHandler()(request, h)
+    }
+  }
+
+  protected getPayload(request: Request) {
+    const payload = (request.payload || {}) as FormPayload
+    const name = this.getComponentName()
+    const value = payload[name]
+    const validateResult = joi
+      .array()
+      .items(uploadIdSchema)
+      .single()
+      .default([])
+      .validate(value)
+
+    if (!validateResult.error) {
+      const files = (request.app.files || []) as FilesState
+      payload[name] = validateResult.value.map((str: string) => {
+        return files.find((file) => file.uploadId === str)
+      })
+      payload[name].formAction = files.formAction
+    }
+
+    return payload
+  }
+
+  getErrors(
+    validationResult?: Pick<ValidationResult, 'error'>
+  ): FormSubmissionErrors | undefined {
+    if (validationResult?.error) {
+      const errorList: FormSubmissionError[] = []
+      const arraySizeErrorTypes = ['array.min', 'array.max', 'array.length']
+      const componentName = this.getComponentName()
+
+      validationResult.error.details.forEach((err) => {
+        const type = err.type
+        const path = err.path.join('.')
+        const formatter = (name: string | number, index: number) =>
+          index > 0 ? `__${name}` : name
+        const name = err.path.map(formatter).join('')
+        const href = `#${name}`
+        const lastPath = err.path[err.path.length - 1]
+
+        if (path.startsWith(componentName)) {
+          if (type === 'any.required' && path === name) {
+            errorList.push({ path, href, name, text: err.message })
+          } else if (type === 'any.only' && lastPath === 'fileStatus') {
+            if (err.context?.value === 'pending') {
+              const text = 'The selected file has not fully uploaded'
+              errorList.push({ path, href, name, text })
+            }
+          } else if (type === 'object.unknown' && lastPath === 'errorMessage') {
+            const text = err.context?.value as string
+            errorList.push({ path, href, name, text })
+          } else if (arraySizeErrorTypes.includes(type)) {
+            errorList.push({ path, href, name, text: err.message })
+          }
+        } else {
+          // The error is for another field on the
+          // page so defer to the standard getError
+          errorList.push(this.getError(err))
+        }
+      })
+
+      return {
+        titleText: this.errorSummaryTitle,
+        errorList
+      }
+    }
+  }
+
+  getViewModel(
+    payload: FormPayload,
+    errors?: FormSubmissionErrors
+  ): FileUploadPageViewModel {
+    const viewModel = super.getViewModel(
+      payload,
+      errors
+    ) as FileUploadPageViewModel
+
+    const fileUploadComponent = viewModel.components.find(
+      (component) => component.model.id === this.fileUploadComponent.name
+    ) as FormComponentViewModel
```

Comment:
Not possible for the component not to be found here. We wouldn't get through the constructor if it wasn't present.

However, I'm happy to change it to something but of the two, I prefer mine currently. It better shows the intent of the code, which is to find the component that corresponds (by id/name) to the one initially found in the constructor. Matching on type doesn't express that as well for me. Also, what if two components of that type were found - the first found would be returned (which might not be the correct one).

Reply:
I'm not sure we can get the compiler to be happy without the `as` casting tbh. It can't narrow:

```
export type ComponentCollectionViewModel = (
  | FormComponentViewModel
  | ContentComponentViewModel
)[]
```

to `FormComponentViewModel` regardless of the predicate we use in `find`.